### PR TITLE
[Log Collector] Fix lock race conditions and state overwrites

### PR DESCRIPTION
**Pull request by TomerShor**
This PR introduces several fixes to the log collector server:

- The 2 state stores that the sever has were renamed for better readability and usability:
  - `stateManifest` - this is the persisted state and it is the source of truth. It depicts which runs should be log collected, and is periodically written to a file for persisting the state in case of pod restarts.
  - `currentState` - this state holds the runs that currently have running goroutines collecting logs. It is used for monitoring by going through the runs in the state manifest and verifying they are present in the current state.
- The file state store client has 2 mutex locks - one for the state itself, and one for reading/writing the persisted file. 
There were some race conditions between those two locks that have sometime caused a deadlock, making the log collector server stuck. These race condition have now been fixed.
- The file state store implemented `GetItemsInProgress` by reading from the file. 
The state manifest was written to the file in a periodic background task.
This caused some runs to be discarded form the state manifest, in the case that the state was overwritten with the old state before the new state manifest was written to the file.
Hence, now the monitoring is using the state manifest map and only reading from the file during initialization.
- In the monitoring goroutine, start logs for runs in parallel for faster handling.

Fixes https://jira.iguazeng.com/browse/ML-3460

### CODE REVIEW

File: go/pkg/services/logcollector/server.go [hunk 1]
```diff
@@ -37,25 +37,33 @@ import (
...
-	baseDir                      string
-	kubeClientSet                kubernetes.Interface
-	stateStore                   statestore.StateStore
-	inMemoryState                statestore.StateStore
+	namespace     string
+	baseDir       string
+	kubeClientSet kubernetes.Interface
+	isChief       bool
+
+	// the state manifest determines which runs' logs should being collected, and is persisted to a file
```
Comment (quaark) [inline go/pkg/services/logcollector/server.go:53]:
```suggestion
	// the state manifest determines which runs' logs should be collected, and is persisted to a file
```

File: go/pkg/services/logcollector/statestore/file/client.go [hunk 2]
```diff
@@ -54,19 +54,27 @@ func NewFileStore(logger logger.Logger, baseDirPath string, stateFileUpdateInter
 func (s *Store) Initialize(ctx context.Context) error {
 	var err error
 
+	// load state from file before starting the update loop, as thr file is our source of truth
```
Comment (quaark) [inline go/pkg/services/logcollector/statestore/file/client.go:57]:
```suggestion
	// load state from file before starting the update loop, as the file is our source of truth
```

File: go/pkg/services/logcollector/statestore/file/client.go [hunk 3]
```diff
@@ -54,19 +54,27 @@ func NewFileStore(logger logger.Logger, baseDirPath string, stateFileUpdateInter
...
 	}
 
+	// we do not defer unlock because we need the lock for writing the state to file
+	s.fileLock.Unlock()
+
+	// write state to file to make sure it exists
+	if err := s.writeStateToFile(s.State); err != nil {
+		return errors.Wrap(err, "Failed to write state file")
+	}
+
```
Comment (quaark) [inline go/pkg/services/logcollector/statestore/file/client.go:72]: This is an unsafe read-write you've implemented here. A is source of truth. go-routine-1 reads A, go-routine-2 reads A, go-routine-1 writes B, go-routine-2 writes C - B is lost.

You need to remove the use of the lock from the `s.writeStateToFile` and have one lock to encapsulate both read and write to one commit.

**Review by quaark — CHANGES_REQUESTED** [review]
There's an issue in how you've implemented the lock fix you should check out. The rest of the changes are 🔥

**Review by quaark — APPROVED** [review]
🚀

**Review by liranbg — APPROVED** [review]

**Pull request merged**